added api management guideline (#7424)
eclipse-theia/theia: doc/api-management.md

Usually, the version management is built around API visibility.
Particularly, if a public API is broken then the major release is required.

Following the conventional approach is not practicable for Theia since all APIs are more or less public.
It will slow down API innovation, accumulate technical debt or require many major releases.

Instead, we require a major release only if a stable API is broken.
Breaking an experimental API can be done with a minor release.

## Stability

Conceptually, an API is all assumptions adopters depend on to make use or extend Theia.
A stable API is based on assumptions that are not subject to change.
It does not matter whether an API is public or internal, or whether it is used a lot or never.

> For instance, the language server protocol depends on such data types like URIs and positions
since they don't change from language to language. It ensures its stability.

API stability is indicated explicitly by adding `@experimental` or `@stable` js-doc tags.
An API without a stability tag is considered to be experimental.

The explicit stability tag should mention since which version an API is stable or experimental.

```ts
/**

>  * @since 0.1.0 - added in 0.1.0

Reviewer: The `since` tag is not mentioned in the text. Would you propose to transform `experimental` into `since` upon finalization?

Author: > The since tag is not mentioned in the text.

I've just wanted to be explicit here. But if it is obvious then I can remove it.

The thing is I don't want to bother anyone with annotations while they adding new APIs except the case when they touch stable APIs already. Otherwise it increase friction on PR reviews which is already quite high.

> Would you propose to transform experimental into since upon finalization?

Not sure about transforming but adding since tag to stable API sounds like a good idea. I will add it.

Author: Another pro for since tag is linting. If we want to have something like PDE api tools it would help instead of using git and github API, but we can add it later too.

>  * @stable since 1.0.0 - finalized in 1.0.0

Reviewer: `@stable since 1.0.0` and `finalized in 1.0.0` are redundant. Do we need any more explanation than `@stable since 1.0.0`?

Author: I will remove it, wanted to explicit

- Experimental APIs don't follow [semver](https://semver.org/#spec-item-8) semantic.
- Experimental APIs can be changed or removed without [the deprecation cycle](#deprecation) if they were not widely adopted.
- Adoption should be measured by the number of internal clients or based on the feedback of Theia contributors and committers.

### Stable

- [Experimental APIs](#experimental) can be graduated to `stable` via [the finalization cycle](#finalization).
- Stable APIs should be based on design decisions that are not likely to change.
- Stable APIs should have sufficient adoption.
- Stable APIs should have proper documentation.
- Stable APIs must follow [semver](https://semver.org/#spec-item-8) semantic.
- They can be changed only in a backward-compatible fashion.
- They can be removed only via [the deprecation cycle](#deprecation).

## Finalization

API finalization should be requested via GitHub issue with `api-finalization` label.

Finalization implies a review of the adoption, stability, and documentation of APIs.
One cannot judge stability without gaining enough experience
and provide stable documentation without API stability in the first place.

> If there is no enough adoption then finalization should be postponed.

Reviewer: ```suggestion
If there is not enough adoption then finalization should be postponed.
```

If it does not get adopted later, it should be considered to move APIs

> from the framework to requesting adopters.

Reviewer: ```suggestion
from the framework to another repository managed by requesting adopters.
```

If APIs are based on design decisions which are subject to change then
one could postpone finalization before such decisions are resolved.
If it is impossible then APIs should be refactored to hide such assumptions.

If sufficient adoption and API stability are established,
documentation should be reviewed and completed.

## Deprecation

API deprecation is indicated by adding `@deprecated` js-doc tag.

> The depreciation tag should mention since which version is deprecated, explain why and what should be used instead.

Reviewer: ```suggestion
The deprecation tag should mention since which version it is deprecated, explain why and what should be used instead.
```

# API Management

> > Guidance below for Theia framework. End products will get better development experience by using private visibility.

Reviewer 2: ```suggestion
> The guidance below is for the code of the Theia framework only. End products will get better development experience by using private visibility.
```

> > Guidance below for Theia framework. End products will get better development experience by using private visibility.

Reviewer 3: `for Theia` -> `for the Theia`

- [**Stability**](#stability)
  - [**Experimental**](#experimental)
  - [**Stable**](#stable)
- [**Finalization**](#finalization)
- [**Deprecation**](#deprecation)

Theia is a framework embracing openness, extensibility, and customizability as much as possible:

> - API defaults to public for clients.

Reviewer 3: `public` -> `public visibilty`

- API defaults to protected for extenders.

> - Language constructions prohibiting runtime access to internals are never used and so on.

Reviewer 2: ```suggestion
- Language constructions prohibiting runtime access to internals are never used.
```

Usually, the version management is built around API visibility.

> Particularly, if a public API is broken then the major release is required.

Reviewer 2: ```suggestion
Particularly, when a public API is broken a new major release is required.
```

> Particularly, if a public API is broken then the major release is required.

Reviewer 3: `is broken then the major release is required.` -> `is broken then, a new major release is required.`

> Following the conventional approach is not practicable for Theia since all APIs are more or less public.

Reviewer 3: `Since all APIs are more or less public, following the conventional approach is not practicable for Theia.`

It will slow down API innovation, accumulate technical debt or require many major releases.

> Instead, we require a major release only if a stable API is broken.

Reviewer 2: ```suggestion
Instead, we demand a major release only if a _stable_ API is broken.
```

> Breaking an experimental API can be done with a minor release.

Reviewer 2: ```suggestion
Breaking an _experimental_ API is allowed in a minor release.
```

## Stability

> Conceptually, an API is all assumptions adopters depend on to make use or extend Theia.

Reviewer 2: ```suggestion
Conceptually, an API consists of all assumptions adopters depend on to make use or extend Theia.
```

A stable API is based on assumptions that are not subject to change.
It does not matter whether an API is public or internal, or whether it is used a lot or never.

> > For instance, the language server protocol depends on such data types like URIs and positions

Reviewer 3: `the language server protocol` -> `the language server protocol (LSP)`, since you are using the LSP abbreviation below.

since they don't change from language to language. It ensures its stability.

API stability is indicated explicitly by adding `@experimental` or `@stable` js-doc tags.

The explicit stability tag should be accompanied by `@since` tag indicating when API was added.
The `@stable` tag should mention since which version an API was finalized.

An API without a stability tag is considered to be experimental and does not require `@since` tag.

```ts
/**
 * One does not need any annotations while working on experimental APIs.
 */
export interface ExperimentalInterface {
}

/**
 * @since 0.1.0
 * @stable since 1.0.0
 */
export interface StableInterface {

    /**
     * The same as `StableInterface`.
     */
    stableMethod(): void;

    /**
     * Adding new API to stable API should be explicit.
     * 
     * @since 1.1.0
     * @experimental
     */
    experimentalMethod(): void;

}

```

### Experimental

> - All new APIs should always be added as **experimental** since it's almost impossible to get [stable API](#stable) right first time.

Reviewer 3: `right first time.` -> `right the first time.`

- Experimental APIs don't require the stability tag, but if a new member is added to [stable API](#stable) then it should be explicitly annotated.
- Experimental APIs don't require extensive documentation. It does not mean that one should not document not obvious parts.
- Experimental APIs don't follow [semver](https://semver.org/#spec-item-8) semantic.

> - Experimental APIs can be changed or removed without [the deprecation cycle](#deprecation) if they were not widely adopted.

Reviewer 3: `can` -> `could`

- Adoption should be measured by the number of internal clients or based on the feedback of Theia contributors and committers.

### Stable

- [Experimental APIs](#experimental) can be graduated to `stable` via [the finalization cycle](#finalization).
- Stable APIs should be based on design decisions that are not likely to change.
- Stable APIs should have sufficient adoption.
- Stable APIs should have proper documentation.
- Stable APIs must follow [semver](https://semver.org/#spec-item-8) semantic.
- They can be changed only in a backward-compatible fashion.
- They can be removed only via [the deprecation cycle](#deprecation).

## Finalization

> API finalization should be requested via GitHub issue with `api-finalization` label.

Reviewer 3: `via GitHub issue` -> `via a GitHub issue` (or `via GitHub issues`)

Finalization implies a review of the adoption, stability, and documentation of APIs.
One cannot judge stability without gaining enough experience
and provide stable documentation without API stability in the first place.

> If there is not enough adoption then finalization should be postponed.

Reviewer 3: `If there is not enough adoption then` -> `If there is not enough adoption, then`

If it does not get adopted later, it should be considered to move APIs
from the framework to another repository managed by requesting adopters.

> Why? Consider the case if the LSP would have a feature that can be supported only by one tool.
It is reasonable to move such a feature to a tool specific LSP extension
instead of bloating the protocol. The same applies to the framework:
If API is added and used only by one adopter, then such API has to be moved to the adopter.
It ensures control of API for such adopters and reduces the API surface of the framework.

> If APIs are based on design decisions which are subject to change then

Reviewer 3: `which are subject to change then` -> `that are subject to change, then`

one could postpone finalization before such decisions are resolved.

> Why? Some design decisions are only temporarily subject to change, for instance, 
which layout framework should be used. After it is resolved, 

> such a design decision becomes fundamental, cannot be changed and can be adopted.

Reviewer 3: `cannot be changed and can be adopted.` -> `cannot be changed, and can be adopted.`

> If it is impossible then APIs should be refactored to hide such assumptions.

Reviewer 3: `If it is impossible then APIs` -> `If it is not possible, then APIs`

> > Why? Some design decisions are changeable by nature, for instance for the LSP, 

Reviewer 3: `, for instance for` -> `, for instance, for`

kind of symbols of a concrete language. In such a case, 
API should be changed to operate abstract symbol data type to hide concrete symbols.

If sufficient adoption and API stability are established,
documentation should be reviewed and completed.

## Deprecation

API deprecation is indicated by adding `@deprecated` js-doc tag.

> The deprecation tag should mention since which version it is deprecated, explain why and what should be used instead.

Reviewer 3: `should mention since which version` -> `should mention, since which version`

```ts
/**
 * @since 0.1.0
 * @stable since 1.0.0
 * @deprecated since 1.1.0 - because that and that, use that instead
 */
export interface DeprecatedStableInterface {

}
```

> Deprecated [stable API](#stable) can be removed after at least one *major* release.

Reviewer 2: ```suggestion
Deprecated [stable API](#stable) can be removed in one of the next *major* releases.
```

Reviewer 2: Or do you mean it has to be deprecated for an entire major release?

Author: No, i meant usual semantic, that it can be removed with next major release.

> Deprecated [experimental API](#experimental) can be removed after at least one *minor* release.

Reviewer 2: Same as above

# API Management

> Guidance below for Theia framework. End products will get better development experience by using private visibility.

Theia is a framework embracing openness, extensibility, and customizability as much as possible:
- API defaults to public for clients.
- API defaults to protected for extenders.
- Language constructions prohibiting runtime access to internals are never used and so on.

Usually, the version management is built around API visibility.
Particularly, if a public API is broken then the major release is required.

Following the conventional approach is not practicable for Theia since all APIs are more or less public.
It will slow down API innovation, accumulate technical debt or require many major releases.

Instead, we require the major release only if a stable API is broken.

> Breaking an experimental API can be done with the minor release,

Reviewer: ```suggestion
Instead, we require a major release only if a stable API is broken.
Breaking an experimental API can be done with a minor release,
```

> but not necessary if such API does not have sufficient adoption.

Reviewer: I don't understand this part. What does "necessary" refer to?

Author: i meant if there are no adopters then it can be just removed. I will remove it, it is explained below.

## Stability

Conceptually, an API is all assumptions adopters depend on to make use or extend Theia.
A stable API is based on assumptions that are not subject to change.
It does not matter whether an API is public or internal, or whether it is used a lot or never.

> For instance, the language server protocol depends on such data types like URIs and positions
since they don't change from language to language. It ensures its stability.

The API stability indicated by adding, exclusively, `@experimental` or `@stable` js-doc tags.

> An API without a stability tag considered to be experimental.

Reviewer: ```suggestion
API stability is indicated explicitly by adding `@experimental` or `@stable` js-doc tags.
An API without a stability tag is considered to be experimental.
```

### Experimental

- All new APIs should always be added as `experimental`
since it's almost impossible to get stable API right first time.
- Usually, experimental APIs don't require the stability tag,
but if a new member is added to `stable` API then it should be explicitly annotated.

> - Such APIs should not have extensive documentation to avoid the maintenance burden.

Reviewer 3: `Such APIs should not have extensive documentation`

I personally do not agree with this. Can we leave this out?

Imagine the following, someone introduces a new API, which is experimental, so it is not documented extensively. Then after some time, the API becomes stable; will you or the author of the API go back and do nothing but write documentation? It sounds good but it is far from reality.

Author: My point is that noone should expect that docs are up-to-date, asks during reviews and so on. If they want to have such level they should ask to make such API stable, then we will care about proper docs and backward compatibility as per semver. If some concrete developer wants to put effort initially it is up to him. I will try to rephrase it.

Reviewer: That's true. Avoiding maintenance burden should not be an excuse for not writing documentation. IMHO good documentation is necessary if we want to see this framework more widely adopted.

We could approach it like this.

 * For experimental API, the amount of documentation should be well balanced: document the parts that are not obvious, but be careful where changes might be expected in the future.
 * Graduation of an API to stable involves a review of its documentation.

Author: > Then after some time, the API becomes stable; will you or the author of the API go back and do nothing but write documentation? It sounds good but it is far from reality.

That's a requirement for stable API. There is a GH  issue to make some API stable. We take time to look at API again, make sure that it is not going to be changed next year and document it. If it was not done API does not become stable.

> - They can be changed or removed without the deprecation cycle if it was not widely adopted.

Reviewer: ```suggestion
- They can be changed or removed without the deprecation cycle if they were not widely adopted.
```

- Adoption should be measured by the number of internal clients or based on the feedback of Theia contributors and committers.

### Stable

> - Experimental APIs can be graduated to `stable` after at least one major release.

Reviewer 4: Just to confirm: This means that a newly added experimental API cannot be made stable until we step the major version? e.g. a new API in v1.x can only be made stable starting at v2.0.0?

Author: Yes, do you know how it could be phrased better?

The major version will mean that we either:
- have experimental APIs with sufficient adoption and not likely to change, so they are matured to be stable
- or we broke stable APIs. Ideally it should never happen if we graduate APIs properly.

Reviewer 5: Why this limit? With Semver adding new API in minor versions is usually ok as long as we don't break existing, no?

Author: > Why this limit? With Semver adding new API in minor versions is usually ok as long as we don't break existing, no?

The approach is to work around semver, i.e. we will follow it only for stable APIs. It won't restrict us to break new APIs without major version.

If someone wants to enforce some API to be stable and follow semver, they will need to ask for the major release with GH issue to annotate and document such APIs. I think it is reasonable since it is extra effort on our side:
- to analyze, stabilize and documents APIs
- to maintain their backward compatibility afterwards

We could decide then for a new major release if we have enough of such API requests.

I could also change it to something like `at least 3 minor releases` instead of major. Or maybe simply that there should be explicit API request via a GH issue.

Reviewer 2: We have to consider how the frequency of major releases affect adopters: While we would embrace stabilizing API (often major releases) each of these major release also sends a signal that API has bean broken.

I'd like the "3 minor releases" much better in this regard

Author: ok, it is true that many have already assumptions about major release semantic, so i will remove it and replace with a requirement for explicit stable API request as GH issue, it also aligns well with https://github.com/eclipse-theia/theia/pull/7424#discussion_r398389145 and https://github.com/eclipse-theia/theia/pull/7424#discussion_r398390178

Timing wise it will be decided from case to case.

- Such experimental APIs must not be based on design decisions that will likely change

> and must have sufficient adoption.

Reviewer: ```suggestion
- Such experimental APIs must not be based on design decisions that will likely change,
and they must have sufficient adoption.
```

- Stable APIs should have proper documentation.
Low maintenance effort is ensured by their stability.
- They can be changed only in a backward-compatible fashion.
- They can be removed only via the deprecation cycle.

## Deprecation

> Deprecated *stable* API can be removed after at least one *major* release.

Reviewer 2: How do we mark deprecation in the code? Another jsdoc annotation?

Author: I thought it is obvious to use `@deprecated` js-doc tag :) but i will make is explicit.

> Deprecated *experimental* API can be removed after at least one *minor* release.

Reviewer 2: We could also mention the targeted removal version in the deprecation annotation

Author: Sounds good, something like
```
@deprecated since 1.1.0 - because that and that, use that instead
```

Reviewer 6: Are deprecated APIs meant to be removed at some point though?

Author: @marechal-p sorry did not notice your comment, yes after release is done, it can be removed in any following releases by that time there was already notice in CHANGELOG.

Do you think i should put it explicit? I don't think we can add target version, since we don't really know. Also it could be good opportunity for first time contributors 😄

Reviewer 6: > I don't think we can add target version, since we don't really know.

Up to you, but if like you said we remove them in next major releases, then we actually know :)
